x-pack/plugins/integration_assistant/server/graphs/log_type_detection/csv.ts

```diff
@@ -0,0 +1,155 @@
+export function upperBoundForColumnCount(csvSamples: string[]): number {
+  return Math.max(0, ...csvSamples.map((sample) => sample.split(',').length));
+}
+
+// Generates a list of temporary column names.
+export function generateColumnNames(count: number): string[] {
+  return Array.from({ length: count }).map((_, i) => `column${i + 1}`);
+}
+
+// Converts a column name into a safe one to use in the `if ctx...` clause.
+// Result must pass rules at https://www.elastic.co/guide/en/elasticsearch/painless/8.15/painless-identifiers.html
+export function toSafeColumnName(columnName: unknown): string | undefined {
+  if (typeof columnName === 'number') {
+    return `Column${columnName}`;
+  }
+  if (typeof columnName !== 'string') {
+    return undefined;
+  }
+  if (columnName.length === 0) {
+    return undefined;
+  }
+  const safeName = columnName.replace(/[^a-zA-Z0-9_]/g, '_');
+  return /^[0-9]/.test(safeName) ? `Column${safeName}` : safeName;
+}
+
+// Returns the column list from a header row. We skip values that are not strings.
+export function columnsFromHeader(
+  tempColumnNames: string[],
+  headerObject: { [key: string]: unknown }
+): Array<string | undefined> {
+  const maxIndex = tempColumnNames.findLastIndex(
+    (columnName) => headerObject[columnName] !== undefined
+  );
+  return tempColumnNames
+    .slice(0, maxIndex + 1)
+    .map((columnName) => headerObject[columnName])
+    .map(toSafeColumnName);
+}
+
+// Count the number of columns actually present in the rows.
+export function totalColumnCount(
+  tempColumnNames: string[],
+  csvRows: Array<{ [key: string]: unknown }>
+): number {
+  return (
+    Math.max(
+      -1,
+      ...csvRows.map((row) =>
+        tempColumnNames.findLastIndex((columnName) => row[columnName] !== undefined)
+      )
+    ) + 1
+  );
+}
+
+// Prefixes each column with the provided prefixes, separated by a period.
+function prefixColumns(columns: string[], prefixes: string[]): string[] {
+  return columns.map((column) => [...prefixes, column].join('.'));
+}
+
+// Converts CSV samples into JSON samples.
+export async function convertCSVSamples({
+  state,
+  client,
+}: LogDetectionNodeParams): Promise<Partial<LogFormatDetectionState>> {
+  const packageName = state.packageName;
+  const dataStreamName = state.dataStreamName;
+
+  const samples = state.logSamples;
+  const temporaryColumns = generateColumnNames(
+    Math.min(upperBoundForColumnCount(samples), MAX_CSV_COLUMNS)
+  );
+  const temporaryProcessor = createCSVProcessor('message', temporaryColumns);
+
+  const { pipelineResults: tempResults, errors: tempErrors } = await createJSONInput(
+    [temporaryProcessor],
+    samples,
+    client
+  );
+
+  if (tempErrors.length > 0) {
+    const exampleError = JSON.stringify(tempErrors[0]);
+    throw new Error(
+      `Failed to convert CSV samples to JSON using the csv processor (${tempErrors.length} errors encountered, such as ${exampleError})`
+    );
+  }
```

Comment:
Why shouldn't the user receive the original error by the Elasticsearch? This contains important information about what is wrong with the CSV.

Reply:
See the Potsgres example above, it tells the user about the unmatched quote. Before I looked into the error (using `console.log`, and later implementing the message) I was confused about why the CSV was not accepted.

Reply:
![image](https://github.com/user-attachments/assets/b53c8679-f8bc-4f23-81dc-7ce6d3dbe578)
This does not look good does it? We can still inform user that there is a problem in CSV format by mapping it to a user readable error and may be for future log the original error.

Reply:
Also what is the character limit on the user facing error? If the error is too long then it does not get displayed fully to user anyway.

Reply:
If this does not look good we can definitely ask for a design help, but let me try to understand the change you are requesting in this PR: it's to only show 

> Bad Request(400): Failed to convert ...

instead of 

> Bad Request(400): Failed to convert ... (such as ... explanation)

?

Reply:
Actually the whole string `Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}} ` is under our control, so I'll just change it.

Reply:
The new message is 

<img width="611" alt="image" src="https://github.com/user-attachments/assets/29721451-e0f3-4496-95de-53c1ac3cbdd8">

Reply:
Looks good. But the user soes not need to know that csv processor is converting the samples into JSON object, right?

User has to simply know: Failed to  analyze the samples due to: unmatched Quote. Please verify the samples.

Reply:
The reason I have the `csv processor` is to underline that we're using the Elasticsearch's processor and not some other library, because it is possible that another library is able to parse the CSV, but not the logic we have in the CSV processor. 

Ideally we would link to the definition of the CSV processor to make clear what we are talking about. I agree otherwise some people would not know the definition. One other wording I considered is "Elasticsearch failed to convert samples to JSON". Prrhaps that?

Reply:
I think we are not posing out any processor errors out to the user. May be should we try and find the samples errors earlier? And error out with 'samples error'. 

And I think Automatic Import does not have to error out Elasticsearch processor errors , since it is 'testpipeline' that we execute and that's more of an internal validation of the processor and not necessarily product error.

Reply:
It's the test pipeline, but it will become part of the actual pipeline we generate, so if there are errors with the sample data then there will be the same errors with the actual data. At least until now, we avoided creating integrations that are known to produce errors!

> we are not posing out any processor errors out to the user.

We generally try to fix the errors by changing the generated pipeline, such as changing the grok format. However, the CSV pipeline does not have any parameters (we could theoretically also consider other separators, like `\t`, but they don't seem to exist among our integrations). If the processor can't process the data – that's it, we can't create a good integration. The user in this case is genuinelly the only one who can proceed:

1. Maybe the sample data got corrupted – then they can be fixed.
2. Maybe it's okay to produce an error for a specific malformed sample – then this sample can be removed.
3. Maybe the format is not an actual CSV, so what we have right now can't generate an integration for it – this whole idea has to be abandoned. This is what happens with Postgres.

In the latter case we can discuss adapting to specific CSV issues. The actual Postgres integration, for example, uses this agent setting:

```
multiline:
  pattern: '^\d{4}-\d{2}-\d{2} '
  negate: true
  match: after
``` 

It's not terribly hard to fix this specific issue (LLM would be asked to find the multiline pattern, and we'll have it saved in the `samplesFormat`) if we want, but we also can't fix all issues like that.

Reply:
I do agree the wording might make people think that someone just quickly bubbled up some low-level error without much though; this would be unfortunately after so much discussion. Perhaps

> Cannot parse the file as the CSV data (reason: ...). Please check the provided samples.

Reply:
Yeah. A more sample related error would be appropriate IMO instead of talking about internals related to processors and test pipeline. 

All user needs to know is that `The samples are not valid and the reason being...xxxxx`. User does not need to know how we found this out , IMO.

I would back this

>Cannot parse the samples as the CSV data (reason: ...). Please check the provided samples.